rebasebot/lifecycle_hooks.py

```diff
@@ -0,0 +1,171 @@
+#    Copyright 2024 Red Hat, Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""This module manages user provided scripts that are executed during the rebase process."""
+
+import logging
+import os
+import select
+import shutil
+import subprocess
+import sys
+import tempfile
+from enum import Enum
+
+import git
+
+
+class LifecycleHook(Enum):
+    """LifecycleHook is an enum of points of the rebase process where scripts can be attached."""
+    PRE_REBASE = "preRebaseHook"
+    PRE_CARRY_COMMIT = "preCarryCommitHook"
+    POST_REBASE = "postRebaseHook"
+    PRE_PUSH_REBASE_BRANCH = "prePushRebaseBranchHook"
+    PRE_CREATE_PR = "preCreatePRHook"
+
+
+class LifecycleHookScript:
+    """LifecycleHookScript represents a script file that can be executed."""
+    script_location: str  # The user-defined location of the script
+    script_file_path: str  # The resolved absolute path to the script
+
+    def __init__(self, script_location: str):
+        self.script_location = script_location
+        if script_location.startswith("git:"):
+            return
+
+        # Replace [BUILTIN] with the absolute path to the builtin scripts directory
+        builtin_hooks_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "builtin-hooks")
+        script_file_path = script_location.replace("[BUILTIN]", builtin_hooks_path)
+        # Save absolute path as the working directory changes during execution
+        script_file_path = os.path.abspath(script_file_path)
+        if not os.path.exists(script_file_path):
+            raise ValueError(f"Script file {script_file_path} does not exist")
+        self.script_file_path = script_file_path
+
+    def fetch_from_git(self, gitwd: git.Repo, temp_hook_dir: str):
+        """Fetches the script from a git repository and stores it in a temporary directory."""
+        # Checks if the script path references a git branch
+        # The format is git:<git_reference>:<file_path>
+        if self.script_location.startswith("git:"):
+            ref, file_path = self.script_location[4:].split(":", 1)
+            hash_suffix = str(abs(hash(self.script_location)))[:5]  # Hash the script location to avoid name conflicts
```

Comment:
Is it always guaranteed that the directory level depth of where the script is located is at :4 and :5 ?

Reply:
No, I think you are reading it wrong.

self.script_location[4:] cuts off the `git:`
str(abs(hash(self.script_location)))[:5] creates an integer hash, converts it into string and takes 5 digits from it.

The full path is not recreated for the temporary script directory and to avoid name collision I added the 5 digit hash to the end of the file.

git:dest/main:rebasebot/script.sh will create file under /tmp/rebasebot_temp/script-15674.sh
git:dest/main:different/path/to/script.sh will create file under /tmp/rebasebot_temp/script-89763.sh

Reply:
Ok makes sense, would be useful to add this as a comment for future readers.

Reply:
Added comment to explain why the hash is there.
And added `git_location = self.script_location[4:]` to make it more obvious that it is cutting off the `git:` prefix.

Reply:
Ty